webrender/src/device/gfx/descriptor_set.rs

```diff
@@ -0,0 +1,167 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use hal::pso::{DescriptorSetLayoutBinding, DescriptorType as DT, ShaderStageFlags as SSF};
+use super::super::{ShaderKind, VertexArrayKind};
+
+macro_rules! descriptor_set_layout_binding {
+    (
+        $binding:expr,
+        $ty:expr,
+        $stage_flags:expr,
+        $immutable_samplers:expr
+    ) => {
+        DescriptorSetLayoutBinding {
+            binding: $binding,
+            ty: $ty,
+            count: 1,
+            stage_flags: $stage_flags,
+            immutable_samplers: $immutable_samplers,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Hash)]
+pub(super) enum DescriptorGroup {
+    Default,
+    Clip,
+    Primitive,
+}
+
+impl From<ShaderKind> for DescriptorGroup {
+    fn from(kind: ShaderKind) -> Self {
+        match kind {
+            ShaderKind::Cache(VertexArrayKind::Border) | ShaderKind::Cache(VertexArrayKind::LineDecoration)
+                | ShaderKind::DebugColor | ShaderKind::DebugFont
+                | ShaderKind::VectorStencil | ShaderKind::VectorCover => DescriptorGroup::Default,
+            ShaderKind::ClipCache => DescriptorGroup::Clip,
+            _ => DescriptorGroup::Primitive,
+        }
+    }
+}
+
+pub(super) const DEFAULT_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
+];
+
+pub(super) const DEFAULT_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0 sampler
+    descriptor_set_layout_binding!(12, DT::Sampler, SSF::ALL, false),
+    // Color1 sampler
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // Dither sampler
+    descriptor_set_layout_binding!(0, DT::Sampler, SSF::ALL, false),
+];
+
+pub(super) const DEFAULT_SET_2: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0
```

Comment:
why are we providing color textures separately from color samplers? If they are meant to be together all the time, we should just go for `CombinedImageSampler`. Or do we know the samplers earlier (and they change less often) than the images?